File: atsa/src/main/java/jp/kaiz/atsassistmod/controller/TrainControllerManager.java
package jp.kaiz.atsassistmod.controller;

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity;
import jp.kaiz.atsassistmod.network.ATSAModNet;
import jp.kaiz.atsassistmod.network.payload.HudPayload;
import jp.kaiz.atsassistmod.rtm.RtmTrains;
import net.minecraft.server.MinecraftServer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Tracks one {@link TrainController} per formation, keyed by the head car's entity
 * id (the old code keyed by {@code Formation.id}, which RTM no longer exposes; see
 * {@code RtmTrains#formationKey}). Driven once per server tick.
 */
public final class TrainControllerManager {
    private static final Map<Long, TrainController> trackingTrainMap = new HashMap<>();

    private TrainControllerManager() {}

    public static TrainController getTrainController(TrainEntity train) {
        if (train == null) {
            return TrainController.NULL;
        }
        long key = RtmTrains.formationKey(train);
        TrainController controller = trackingTrainMap.computeIfAbsent(key, k -> new TrainController(RtmTrains.head(train)));
        controller.bind(RtmTrains.head(train));
        return controller;
    }

    /** Server-side: advance every tracked controller; drop stale ones; sync HUD. */
    public static void onTick(MinecraftServer server) {
        if (trackingTrainMap.isEmpty()) {
            return;
        }
        List<Long> delList = new ArrayList<>();
        trackingTrainMap.forEach((key, controller) -> {
            TrainEntity train = controller.getTrain();
            if (train == null || train.isRemoved() || !RtmTrains.isControlCar(train)
                    || RtmTrains.formationKey(train) != key) {
                delList.add(key);
                return;
            }
            try {
                controller.onUpdate();
            } catch (Exception e) {
                e.printStackTrace();
            }
            ATSAModNet.broadcastHud(server, toHud(key, controller));
        });
        delList.forEach(key -> {
            ATSAModNet.broadcastHud(server, HudPayload.remove(key));
            trackingTrainMap.remove(key);
        });
    }

    private static HudPayload toHud(long key, TrainController c) {
        return new HudPayload(1, key,
                c.isATO(), c.tascController.isEnable(), c.getTrainProtectionType().id,
                c.getATOSpeedLimit(), (int) c.tascController.getStopDistance(),
                c.getSpeedLimit(), c.getTrainProtectionSpeedLimit(), c.isManualDrive());
    }

    public static TrainController find(long key) {
        return trackingTrainMap.get(key);
    }

    public static void clear() {
        trackingTrainMap.clear();
    }
}

File: atsa/src/main/kotlin/jp/kaiz/atsassistmod/controller/TrainControllerManager.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package jp.kaiz.atsassistmod.controller

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity
import jp.kaiz.atsassistmod.network.ATSAModNet
import jp.kaiz.atsassistmod.network.payload.HudPayload
import jp.kaiz.atsassistmod.rtm.RtmTrains
import net.minecraft.server.MinecraftServer

/**
 * Tracks one [TrainController] per formation, keyed by the head car's entity id.
 */
object TrainControllerManager {
    private val trackingTrainMap = HashMap<Long, TrainController>()

    @JvmStatic
    fun getTrainController(train: TrainEntity?): TrainController {
        if (train == null) {
            return TrainController.NULL
        }
        val key = RtmTrains.formationKey(train)
        val controller = trackingTrainMap.computeIfAbsent(key) {
            TrainController(RtmTrains.head(train))
        }
        controller.bind(RtmTrains.head(train))
        return controller
    }

    /** Server-side: advance every tracked controller; drop stale ones; sync HUD. */
    @JvmStatic
    fun onTick(server: MinecraftServer) {
        if (trackingTrainMap.isEmpty()) {
            return
        }
        val deleteList = ArrayList<Long>()
        trackingTrainMap.forEach { (key, controller) ->
            val train = controller.getTrain()
            if (
                train == null ||
                train.isRemoved ||
                !RtmTrains.isControlCar(train) ||
                RtmTrains.formationKey(train) != key
            ) {
                deleteList.add(key)
                return@forEach
            }
            try {
                controller.onUpdate()
            } catch (exception: Exception) {
                exception.printStackTrace()
            }
            ATSAModNet.broadcastHud(server, toHud(key, controller))
        }
        deleteList.forEach { key ->
            ATSAModNet.broadcastHud(server, HudPayload.remove(key))
            trackingTrainMap.remove(key)
        }
    }

    private fun toHud(key: Long, controller: TrainController): HudPayload =
        HudPayload(
            1,
            key,
            controller.isATO(),
            controller.tascController.isEnable(),
            controller.getTrainProtectionType().id,
            controller.getATOSpeedLimit(),
            controller.tascController.getStopDistance().toInt(),
            controller.getSpeedLimit(),
            controller.getTrainProtectionSpeedLimit(),
            controller.isManualDrive(),
        )

    @JvmStatic
    fun find(key: Long): TrainController? = trackingTrainMap[key]

    @JvmStatic
    fun clear() {
        trackingTrainMap.clear()
    }
}
